File: app/backend/hawkeye_backend/models/household.py
"""Who belongs in this house, and what they carry.

A roster entry is a human declaration, never a sensed fact. CSI resolves a body
reflecting RF and cannot recognise a person, so every persistent identity here
comes from a device someone carries or from a resident saying so out loud.

The identifier is stored hashed. A roster is a list of which humans were in a
building and which devices they carry, and that is exactly the file that should
not be useful to whoever steals it.
"""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, Field, field_validator

from hawkeye_backend.models.common import Provenance, Source, utc_now


class MemberKind(StrEnum):
    """Whether someone lives here or visits."""

    RESIDENT = "resident"
    GUEST = "guest"


class KnownDevice(BaseModel):
    """A device the roster recognises.

    `identifier_hash` is HMAC-SHA256 over the observed address under the site
    salt. The raw address is never stored, so a leaked roster does not become a
    device-tracking list for the house it came from.
    """

    device_id: str
    identifier_hash: str = Field(min_length=64, max_length=64)
    fingerprint: str = Field(description="Short and human-readable, for telling two phones apart.")
    label: str | None = None
    added_at: datetime = Field(default_factory=utc_now)
    last_seen_at: datetime | None = None


class HouseholdMember(BaseModel):
    """One person the house is not surprised by."""

    member_id: str
    name: str
    kind: MemberKind
    devices: list[KnownDevice] = []
    added_at: datetime = Field(default_factory=utc_now)
    added_by: Literal["approval", "enrollment"] = Field(
        description=(
            "How they joined. Approving a live detection is a different kind of fact "
            "from being typed in during setup, and the Household list says which."
        )
    )
    provenance: Provenance = Field(
        default_factory=lambda: Provenance(source=Source.USER_INPUT, producer="app/ios"),
        description=(
            "A roster entry is a human declaration and is labelled as one. `USER_INPUT` "
            "maps to `SourceClass.HUMAN`, never to anything measured or inferred, so "
            "`caller` can say 'the resident says this person is expected' and cannot "
            "accidentally say 'the system verified this person'."
        ),
    )

    @property
    def recognisable(self) -> bool:
        """False means present but invisible to the roster.

        A named guest with no device is legal and useful, and the UI must render
        the difference rather than leaving it blank.
        """
        return bool(self.devices)


class ObservedDevice(BaseModel):
    """A device seen associated to the network, hashed on arrival.

    Carries `Provenance` because it is a reading like any other. On the mock
    path that provenance says simulated, and nothing downstream has to guess.
    """

    device_id: str
    identifier_hash: str = Field(min_length=64, max_length=64)
    fingerprint: str
    first_seen_at: datetime = Field(default_factory=utc_now)
    provenance: Provenance


class RememberRequest(BaseModel):
    """Name a person and optionally bind the device that just appeared."""

    name: str
    kind: MemberKind = MemberKind.GUEST
    device_id: str | None = Field(
        default=None,
        description="An unclaimed device to bind, or null for a guest with no phone.",
    )

    @field_validator("name")
    @classmethod
    def _name_is_not_blank(cls, v: str) -> str:
        cleaned = v.strip()
        if not cleaned:
            raise ValueError("a roster entry needs a name")
        return cleaned

File: app/ios/HawkEye/Models/Household.swift
import Foundation

/// Who the house is not surprised by, and what they carry.
///
/// This is the backend's `HouseholdMember`, field for field, checked against
/// `app/backend/schema/household.json`. A roster entry is always a human
/// declaration: either the resident approving a presence that just showed up,
/// or naming a visitor from the notice banner. There is no path that infers a
/// member from CSI alone, and `Kind`/`AddedBy` exist to say so on screen.
///
/// `identifier_hash` is deliberately not modelled here. It is an HMAC digest
/// the backend keeps so a leaked roster does not double as a device-tracking
/// list, and the app has no use for it: a value the UI never holds cannot leak
/// from the UI.
struct HouseholdMember: Codable, Sendable, Hashable, Identifiable {

    /// Whether someone lives here or is a visitor. Two words, and the label is
    /// what the roster prints, not the raw case name.
    enum Kind: String, Codable, Sendable, Hashable, CaseIterable {
        case resident
        case guest

        var label: String {
            switch self {
            case .resident: "Lives here"
            case .guest: "Visitor"
            }
        }
    }

    /// How this member joined the roster. Approving a live detection is a
    /// different kind of fact from naming someone during the remember flow
    /// (there is no separate enrollment screen in this app, but the backend's
    /// enum still distinguishes the two, and the client mirrors it rather than
    /// collapsing it).
    enum AddedBy: String, Codable, Sendable, Hashable, CaseIterable {
        case approval
        case enrollment
    }

    var memberID: String
    var name: String
    var kind: Kind
    var devices: [KnownDevice]
    var addedAt: Date
    var addedBy: AddedBy
    var provenance: Provenance
    /// **Computed on the server**, not here. A named guest with no device is
    /// legal, and this is false in that case: present, but invisible to the
    /// roster next time. The client reads this rather than checking
    /// `devices.isEmpty` itself, for the same reason `Provenance.sourceClass`
    /// is server-computed: one source of truth for a fact that decides what
    /// the UI may claim about the system's ability to recognise someone.
    var isRecognisable: Bool

    var id: String { memberID }

    enum CodingKeys: String, CodingKey {
        case name, kind, devices, provenance
        case memberID = "member_id"
        case addedAt = "added_at"
        case addedBy = "added_by"
        case isRecognisable = "recognisable"
    }
}

/// A device the roster recognises, hanging off a `HouseholdMember`.
struct KnownDevice: Codable, Sendable, Hashable, Identifiable {
    var deviceID: String
    var fingerprint: String
    /// What the resident calls it, e.g. "iPhone". Optional: nothing on the
    /// remember sheet collects this yet.
    var label: String?
    var addedAt: Date
    var lastSeenAt: Date?

    var id: String { deviceID }

    enum CodingKeys: String, CodingKey {
        case fingerprint, label
        case deviceID = "device_id"
        case addedAt = "added_at"
        case lastSeenAt = "last_seen_at"
    }
}

/// A device seen associated to the network that nobody has claimed yet.
///
/// This is the binding candidate the remember sheet offers: "this phone joined
/// just now, is it theirs?" It carries `Provenance` like any other reading, so
/// the mock path's simulated device is labelled as such rather than presented
/// as a real association.
struct ObservedDevice: Codable, Sendable, Hashable, Identifiable {
    var deviceID: String
    var fingerprint: String
    var firstSeenAt: Date
    var provenance: Provenance

    var id: String { deviceID }

    enum CodingKeys: String, CodingKey {
        case fingerprint, provenance
        case deviceID = "device_id"
        case firstSeenAt = "first_seen_at"
    }
}

/// `POST /v1/household/remember` request body.
struct RememberRequest: Codable, Sendable, Hashable {
    var name: String
    var kind: HouseholdMember.Kind
    var deviceID: String?

    enum CodingKeys: String, CodingKey {
        case name, kind
        case deviceID = "device_id"
    }
}
